packages/@react-aria/textfield/src/useTextField.ts

```diff
@@ -28,25 +36,34 @@ export interface TextFieldAria {
   errorMessageProps: HTMLAttributes<HTMLElement>
 }
 
-interface AriaTextFieldOptions extends AriaTextFieldProps {
+type TextFieldIntrinsicElementType = keyof Pick<IntrinsicHTMLElements, 'input' | 'textarea'>;
+
+type TextFieldHTMLElementType = Pick<IntrinsicHTMLElements, TextFieldIntrinsicElementType>[TextFieldIntrinsicElementType];
+
+type TextFieldHTMLAttributesType = Pick<IntrinsicHTMLAttributes, TextFieldIntrinsicElementType>[TextFieldIntrinsicElementType];
+
+interface AriaTextFieldOptions<T extends TextFieldIntrinsicElementType = DefaultElementType> extends AriaTextFieldProps {
   /**
    * The HTML element used to render the input, e.g. 'input', or 'textarea'.
    * It determines whether certain HTML attributes will be included in `inputProps`.
    * For example, [`type`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-type).
    * @default 'input'
    */
-  inputElementType?: ElementType
+  inputElementType?: T
 }
 
+export function useTextField(props: AriaTextFieldOptions<'input'>, ref: RefObject<HTMLInputElement>): TextFieldAria<InputHTMLAttributes<HTMLInputElement>>;
+export function useTextField(props: AriaTextFieldOptions<'textarea'>, ref: RefObject<HTMLTextAreaElement>): TextFieldAria<TextareaHTMLAttributes<HTMLTextAreaElement>>;
+
 /**
  * Provides the behavior and accessibility implementation for a text field.
  * @param props - Props for the text field.
  * @param ref - Ref to the HTML input or textarea element.
  */
 export function useTextField(
-  props: AriaTextFieldOptions,
-  ref: RefObject<HTMLInputElement | HTMLTextAreaElement>
-): TextFieldAria {
+  props: AriaTextFieldOptions<TextFieldIntrinsicElementType>,
+  ref: RefObject<TextFieldHTMLElementType>
+): TextFieldAria<TextFieldHTMLAttributesType> {
```

Comment:
I assume it's because of the default signature that it renders like so in the docs, we really need to update our docs to be able to display the different override signatures, out of scope though
<img width="1107" alt="Screen Shot 2021-09-27 at 10 13 09 AM" src="https://user-images.githubusercontent.com/698229/134938825-37df88b6-5ce6-4255-8de9-6f910e68ba07.png">
<img width="1238" alt="Screen Shot 2021-09-27 at 10 12 59 AM" src="https://user-images.githubusercontent.com/698229/134938827-019c01f4-0a1b-45bd-9b7c-152982027871.png">
<img width="1014" alt="Screen Shot 2021-09-27 at 10 12 40 AM" src="https://user-images.githubusercontent.com/698229/134938830-3380a4ce-2727-4453-a6c0-b791f008a06b.png">
<img width="1073" alt="Screen Shot 2021-09-27 at 10 12 22 AM" src="https://user-images.githubusercontent.com/698229/134938832-13af04fa-7924-4261-afca-52fcf9c31e06.png">
any ideas though on how we can make this easier to read in the meantime though?

Reply:
These are looking better :)